### ISSUE TITLE
Disable next line should disable the next effective line

### ISSUE THREAD

**Issue by andrewmiller1** [eslint/eslint#10933]
**What rule do you want to change?**

eslint-disable-next-line

**Does this change cause the rule to produce more or fewer warnings?**

Unchanged

**How will the change be implemented? (New option, new default behavior, etc.)?**

New default behavior

**Please provide some example code that this change will affect:**

```js
const DEFAULT = {


  // eslint-disable-next-line id-length

  MAX_MSG_PER_TEST: 100,

  
  SAMPLE_MS: 500
};
```

**What does the rule currently do for this code?**

Disable the literal next line. With an id-length of 5 <= length <= 12, MAX_MSG_PER_TEST is invalid.

**What will the rule do after it's changed?**

Disable the next line of code. With an id-length of 5 <= length <= 12, MAX_MSG_PER_TEST is valid.

**Comment by not-an-aardvark** [comment]
Hi, thanks for the proposal.

Could you clarify the use case for this? It's not clear to me why it would be useful for `eslint-disable-next-line` to disable something other than the next line of text. Couldn't you achieve the same effect by removing the blank line before `MAX_MESSAGE_PER_TEST`?

Also, what would the "next effective line" line would mean in this context? For example, if a rule reports an error on the blank line before `MAX_MSG_PER_TEST`, would that report also be disabled? In general, errors can be reported on lines that don't contain any code (e.g. if a line contains trailing whitespace, then a rule might create a report about the existence of trailing whitespace).

**Comment by andrewmiller1** [comment]
Hi @not-an-aardvark,

Good questions.

> Could you clarify the use case for this?

My use case is to help readers scan the code. This proposal will allow eslint's disable functionality to support using the eslint lines-around-comment: ["error", { "afterLineComment": true }] rule and option and also a combination of ideas from Code Complete's Layout and Style and Self-Documenting Code chapters.

The immediate affect would be that users enforcing a blank line after each comment get to use eslint-disable-next-line too.

<details><summary>Some of the background</summary>

It's hard for me to scan and even read through files when comments hug lines of code. It camouflages both the code and the comment into one blob. If not separated, I have to stop whatever I'm doing or thinking just to see what it is. Even worse, I'll most likely forget what it is shortly after and have to go back and read it again. Setting up my files in a way I can quickly scan them is essential for me.

I tried various ways of laying out comments until I found one that felt right. These are a portion of the constraints, at least some of the ones from Code Complete, I'm following simultaneously:
1. The first idea is the advice, "**Set off each comment with at least one blank line**".
2. The second idea is the mention of how "some people use **a blank line both before and after the comment**."
3. The third idea is the advice, "Separate routines with a file clearly."
a. "Separate each routine from other routines with at least two blank lines."
b. "Three blank lines are used in this example **so that the separation between** routines **is more noticeable than the blank lines within** each routine."
4. Commenting Individual Lines - "In good code, the need to comment individual lines of code is rare. Here are two possible reasons a line of code would need a comment:
   * The single line is complicated enough to need an explanation.
   * The single line once had an error, and you want a record of the error.
5. Endline Comments and Their Problems
   * They look bad, if not aligned with the other endline comments
   * Tedious to align if using many of them
   * Hard to maintain them all each time one of them grow
6. When to Use Endline Comments
   * "Use endline comments to annotate data declarations"
   * "Use endline comments to mark ends of blocks"
7. "**Justify violations of good programming style**"
8. "Keep comments close to the code they describe" - Another way of saying this: **Keep comments closer to their respective code than to other code.**

<details><summary>Examples (in other languages only to match the examples from Code Complete)</summary>

<caption><b>Code Complete advice</b> Java example of setting off a comment with a blank line.</caption>

```java
// comment zero
CodeStatementZero;
CodeStatementOne;

// comment one
CodeStatementTwo;
CodeStatementThree;
```

<caption><b>Code Complete mention</b> Java example of setting off a comment with two blank lines. (period at the start is to force GitHub to include the line and represent a blank line)</caption>

```java
.
// comment zero

CodeStatementZero;
CodeStatementOne;

// comment one

CodeStatementTwo;
CodeStatementThree;
```

<caption><b>Code Complete advice</b> Visual Basic example of using blank lines between routines.</caption>

```vb
'find the arithmetic maximum of arg1 and arg2
Function Max( arg1 As Integer, arg2 As Integer ) As Integer
   If ( arg1 > arg2 ) Then
      Max = arg1
   Else
      Max = arg2
   End If
End Function



'find the arithmetic minimum of arg1 and arg2
Function Min( arg1 As Integer, arg2 As Integer ) As Integer
   If ( arg1 < arg2 ) Then
      Min = arg1
   Else
      Min = arg2
   End If
end Function
```

<caption><b>Code Complete advice</b> Java Example of Good Endline Comments for Data Declarations</caption>

```java
int boundary = 0;         // upper index of sorted part of array
String insertVal = BLANK; // data elmt to insert in sorted part of array
int insertPos = 0;        // position to insert elmt in sorted part of array
```

</details><br>

The requirement is to be able to use the eslint-disable-next-line function in a declarative manner. The constraint of mine is to have it work, even when there is 1 blank line after a comment, with 2 blank lines before the comment and 2 blank lines after the respective code.

<caption><b>My style</b> JavaScript example of a blank line between the comment and its code, with two blank lines before and after the comment and the code. (periods at the start is to force GitHub to include the lines and represent blank lines)</caption>

```js
.
.
// comment zero

CodeStatementZero;
CodeStatementOne;


// comment one

CodeStatementTwo;
CodeStatementThree;


```

</details><br>

> Couldn't you achieve the same effect by removing the blank line before MAX_MESSAGE_PER_TEST?

Unfortunately no. The effect I'm going for is to have a space between the comment and the code, while also being able to use the the concise option, eslint-disable-next-line.

Currently to get this effect, eslint-disable and eslint-enable must be used. Any spacing requirements are multiplied:

<caption>JavaScript example of one line of code taking up nine lines of the file.</caption>

```js
const DEFAULT = {


  /* eslint-disable id-length */

  MAX_MSG_PER_TEST: 100,

  /* eslint-enable id-length */

  
  SAMPLE_MS: 500
};
```

> Also, what would the "next effective line" line would mean in this context?

Since the rule being disabled isn't about the blank spaces, it's fine for disable to skip checking if the blank spaces meet the id-length rule, until there's actual code to check. Just like you would if you were making the check by hand.

> For example, if a rule reports an error on the blank line before MAX_MSG_PER_TEST, would that report also be disabled?

No. Any event or rule, happening outside the rule context of the eslint-disable-next-line rule instance, should not affect that instance of eslint-disable-next-line rule.

**Comment by andrewmiller1** [comment]
Another option I think would be appropriate is the ability to name exceptions ahead of time. Something like stating at the beginning of the file, or maybe a separate file, to define white-listed labels or indicators. Any variable names or patterns found while evaluating a linting violation, that are also in the white-list, are "blessed" and cause those violations to go unreported.

Unsure if having that would really invalidate the need for this proposal. But it's worth a thought.

**Comment by kaicataldo** [comment]
First of all, thanks for contributing! I'm :-1: to this proposal. While I understand the desire to pad comments with blank newlines for readability's sake, this feels like it would add a lot of complexity (both in behavior and maintainability) to what is currently a very predictable and straightforward behavior. This would also be a breaking change.

If you're worried about the comment being directly above it, have you considered using `eslint-disable-line` line comments at the end of the line in question?

> Since the rule being disabled isn't about the blank spaces, it's fine for disable to skip checking if the blank spaces meet the id-length rule, until there's actual code to check. Just like you would if you were making the check by hand.

> No. Any event or rule, happening outside the rule context of the eslint-disable-next-line rule instance, should not affect that instance of eslint-disable-next-line rule.

Given the above two comments, Ii sounds like making this change would require for rules to have some kind of metadata declaring whether they involve whitespace or not, so that ESLint can know if the whitespace around `eslint-disable-next-line` comments can safely be ignored. This would also require that all plugins with custom rules specify this metadata.

Between it being a breaking change (and one that would potentially require all plugin authors to update their plugins), the increase in complexity of the implementation and maintainability (with less-straightforward behavior), and that there are already two other ways to solve this (that seem to stick to the style you're discussing here, though please correct me if I'm wrong!), I'm not currently convinced this should be included in core.

A new kind of `eslint-disable-*` comment would alleviate some of my concerns, but I think I would need to be convinced of the value of introducing more options when we already have a number of ways to solve this in our current API.

Some related issues:
 - https://github.com/eslint/eslint/issues/5206
 - https://github.com/eslint/eslint/issues/6903

**Comment by andrewmiller1** [comment]
That's understandable. I put my hacker pants on and found I could simply edit ./lib/util/apply-disable-directives.js

from

```js
case "disable-next-line":
    return [
        { type: "disable", line: directive.line + 1, column: 1, ruleId: directive.ruleId, unprocessedDirective: directive },
        { type: "enable", line: directive.line + 2, column: 0, ruleId: directive.ruleId, unprocessedDirective: directive }
    ];
```

to

```js
case "disable-next-line":
    return [
        { type: "disable", line: directive.line + 2, column: 1, ruleId: directive.ruleId, unprocessedDirective: directive },
        { type: "enable", line: directive.line + 3, column: 0, ruleId: directive.ruleId, unprocessedDirective: directive }
    ];
```

I'll also look into language agnostic linting tools to see if I can integrate them into any custom solutions here.

Thanks!
